core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java

```diff
@@ -0,0 +1,180 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.groupby;
+
+import io.questdb.cairo.CairoEngine;
+import io.questdb.cairo.GenericRecordMetadata;
+import io.questdb.cairo.SymbolMapReader;
+import io.questdb.cairo.TableReader;
+import io.questdb.cairo.sql.*;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.std.Misc;
+
+public class DistinctSymbolRecordCursorFactory implements RecordCursorFactory {
+    private final DistinctSymbolRecordCursor cursor;
+    private final CairoEngine engine;
+    private final GenericRecordMetadata metadata;
+    private final CharSequence tableName;
+    private final int columnIndex;
+    private final long tableVersion;
+
+    public DistinctSymbolRecordCursorFactory(
+            final CairoEngine engine,
+            final GenericRecordMetadata metadata,
+            final CharSequence tableName,
+            final int columnIndex,
+            final long tableVersion) {
+        this.engine = engine;
+        this.metadata = metadata;
+        this.tableName = tableName;
+        this.columnIndex = columnIndex;
+        this.tableVersion = tableVersion;
+        this.cursor = new DistinctSymbolRecordCursor();
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) {
+        TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName, tableVersion);
+        CharSequence columnName = reader.getMetadata().getColumnName(columnIndex);
+        int columnIndex = reader.getMetadata().getColumnIndex(columnName);
+        cursor.of(reader, columnIndex);
+        return cursor;
+    }
+
+    @Override
+    public RecordMetadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public boolean isRandomAccessCursor() {
+        return true;
+    }
+
+    private static class DistinctSymbolRecordCursor implements RecordCursor {
+        private DistinctSymbolRecord recordA = new DistinctSymbolRecord();
+        private DistinctSymbolRecord recordB = new DistinctSymbolRecord();
+        private TableReader reader;
+        private SymbolMapReader symbolMapReader;
+        private int numberOfSymbols;
+
+        DistinctSymbolRecordCursor() {
+        }
+
+        @Override
+        public void close() {
+            Misc.free(reader);
```

Comment:
it would be better to assign `reader`, e.g.

```
reader = Misc.free(reader);
```

to defend against double close